packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -433,9 +503,10 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
      * For messages from a remote client, this will be undefined.
      */
     private process(message: ISequencedDocumentMessage, local: boolean, localOpMetadata: unknown) {
-        this.emit("pre-op", message, local, this);
+        this.verifyNotClosed(); // This will result in container closure.
+        this.safeRaiseEvent("pre-op", message, local, this);
         this.processCore(message, local, localOpMetadata);
-        this.emit("op", message, local, this);
+        this.safeRaiseEvent("op", message, local, this);
```

Comment:
A potentially more scoped (and maybe better) option - stop emitting "pre-op" and "op" entirely in `SharedObject`, and only emit in sequence during `processCore` (since most DDS don't need to emit "op")

Reply:
I'm not after these events (yes, they would be nice to be removed, but that seems orthogonal problem).
The problem is all the events (TEvent interface and actual usage in derived classes) and listeners throwing exceptions.

We could get rid of EventEmitterWithErrorHandling here and simply propagating exceptions (not eating them - that's what we do) will do the right thing in most cases (at least when processing remote ops).
My worry with this more simple story is that local changes (and exceptions thrown from events due to local changes) would leave DDS in non-consistent state which is very hard to debug down the road. Basically there is no RAII and it's safe to assume that exception anywhere is a game over.

Reply:
Oh sorry I failed to notice what your super call was doing.

I don't know about closing the container for just any event listener -- event listeners on DDSs (either directly or indirectly through DOs) will include views etc. which based on my experience tend to throw a bunch of silly errors that don't matter.  I worry a bit that overaggressive nuking will make the platform harder to use than is necessary.

You mention in the linked issue that customers should never throw errors to us -- I feel like that is the more correct solution, as catching the error in the handler gives them better recovery options (e.g. kill off a single handler rather than letting the whole container die).  But I do respect the fail-fast strategy, so I wouldn't be strictly opposed.

To minimized confused bug reports on why the container is closed, is there a way to include some text in the `DataProcessingError` like "You threw an error to a SharedObject on event handler 'foo', don't do that, we're closing the container as a result"?  Ideally something that shows up in the error console too I think.

Reply:
Yep, that's the goal - to force customers not to throw at us :) So we need some way to expose such errors to them - silently eating them is not an option IMHO. We can always build some API (containers warnings is that direction, though it has a lot of issues that could be probably improved, if we want to go that route). I prefer fail-fast approach - it helps us find our bugs sooner and it also leaves a way for customers to do right thing.

RE DataProcessingError - I think it should be gone ultimately (in its current form). It should be a property (dataProcessingError: true) on any error (IFluidErrorBase), but it should preserve errorType if it was already IFluidErrorBase error, or add unique errorType (depends on place and condition where its thrown).
Similarly here, it should either preserve errorType (if it's our error) or put unique errorType on this object, but in all cases - add new property (let's call it SharedObjectEventListerException = true).
Basically all these qualities (is it a data processing error, is it DDS listener exception, is it something else) should compose, not overwrite one another.
Text is a bit tricky - we always try to preserve message of original error. There is no error nesting (though maybe we should add it - not for telemetry, but for real-time debugging).

What do you think about such plan?

+@markfields for his critique

Reply:
Yeah, maybe DataProcessingError should instead be something like UserCodeError?  Basically I'm thinking that this error is a little special because it lives on the public API boundary.  So whereas our internal errors are OK with error codes and internal knowledge about error types, this one wants to be very understandable from the outside with user-friendly messaging and visibility.

Plan is prob fine, again it will probably be easier to relax the fail-fast in the future based on feedback rather than make it more strict so starting strict is prob ok.

Reply:
Let's keep this discussion (on how user errors are projected in our system) as part of https://github.com/microsoft/FluidFramework/pull/8475 and appropriate issue. It's not unique to this case as user exceptions come from various points and we need to ensure we are consistent in handling them.
Wrapping (i.e. DataProcessingError.wrapIfUnrecognized()) is the way we will handle it in all cases.